File: sdk/@launchdarkly/observability-android/lib/src/main/kotlin/com/launchdarkly/observability/api/ObservabilityOptions.kt
package com.launchdarkly.observability.api

import com.launchdarkly.observability.BuildConfig
import com.launchdarkly.observability.client.TelemetryInspector
import com.launchdarkly.observability.context.LDObserveLogging
import com.launchdarkly.observability.context.ObserveLogAdapter
import io.opentelemetry.api.common.Attributes
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.minutes

const val DEFAULT_SERVICE_NAME = "observability-android"
const val DEFAULT_OTLP_ENDPOINT = "https://otel.observability.app.launchdarkly.com:4318"
const val DEFAULT_BACKEND_URL = "https://pub.observability.app.launchdarkly.com"

/**
 * Configuration options for the Observability plugin.
 *
 * @property serviceName The service name for the application. Defaults to [DEFAULT_SERVICE_NAME].
 * @property serviceVersion The version of the service. Defaults to the SDK version.
 * @property otlpEndpoint The OTLP exporter endpoint. Defaults to LaunchDarkly endpoint [DEFAULT_OTLP_ENDPOINT].
 * @property backendUrl The backend URL for non-OTLP operations. Defaults to LaunchDarkly url [DEFAULT_BACKEND_URL].
 * @property resourceAttributes Additional resource attributes to include in telemetry data.
 * @property customHeaders Custom headers to include with OTLP exports.
 * @property sessionBackgroundTimeout Session timeout if app is backgrounded. Defaults to 15 minutes.
 * @property debug Enables verbose internal logging if true as well as other debug functionality. Defaults to false.
 * @property logsApiLevel Level for logs to be exported. Defaults to INFO. Set to NONE to disable log exporting.
 * @property tracesApi Options for configuring traces. See [TracesApi]. Tracing is enabled by default.
 * @property metricsApi Options for configuring metrics. See [MetricsApi]. Metrics are enabled by default.
 * @property analytics Options for configuring analytics telemetry. See [Analytics].
 * @property instrumentations Options for configuring automatic instrumentations. See [Instrumentations].
 * @property logAdapter The log adapter to use. Defaults to [LDObserveLogging.adapter] which writes to Android's native Log API.
 * @property loggerName The name of the logger to use. Defaults to "LaunchDarklyObservabilityPlugin".
 * @property telemetryInspector Optional [TelemetryInspector] for intercepting exported telemetry during testing.
 *   When provided together with [debug] = true, the inspector's exporters are wired into composite
 *   exporters so that test code can assert on the data that flows through the SDK.
 */
data class ObservabilityOptions(
    val enabled: Boolean = true,
    val serviceName: String = DEFAULT_SERVICE_NAME,
    val serviceVersion: String = BuildConfig.OBSERVABILITY_SDK_VERSION,
    val otlpEndpoint: String = DEFAULT_OTLP_ENDPOINT,
    val backendUrl: String = DEFAULT_BACKEND_URL,
    val contextFriendlyName: String? = null,
    val resourceAttributes: Attributes = Attributes.empty(),
    val customHeaders: Map<String, String> = emptyMap(),
    val sessionBackgroundTimeout: Duration = 15.minutes,
    val debug: Boolean = false,
    val logsApiLevel: LogLevel = LogLevel.INFO,
    val tracesApi: TracesApi = TracesApi.enabled(),
    val metricsApi: MetricsApi = MetricsApi.enabled(),
    val analytics: Analytics = Analytics(),
    val instrumentations: Instrumentations = Instrumentations(),
    val logAdapter: ObserveLogAdapter = LDObserveLogging.adapter(),
    val loggerName: String = "LaunchDarklyObservabilityPlugin",
    val telemetryInspector: TelemetryInspector? = null,
){
    /**
     * Options for configuring traces.
     *
     * @property includeErrors Whether to automatically instrument for and record errors and exceptions as spans.
     * @property includeSpans Whether to automatically instrument for and record UI performance and other events as spans.
     */
    data class TracesApi(
        val includeErrors: Boolean = true,
        val includeSpans: Boolean = true
    ) {
        /**
         * Java-friendly fluent builder for [TracesApi].
         */
        class Builder {
            private var value = TracesApi()

            fun includeErrors(includeErrors: Boolean) = apply { value = value.copy(includeErrors = includeErrors) }
            fun includeSpans(includeSpans: Boolean) = apply { value = value.copy(includeSpans = includeSpans) }

            fun build() = value
        }

        companion object {
            @JvmStatic
            fun enabled() = TracesApi()

            @JvmStatic
            fun disabled() = TracesApi(includeErrors = false, includeSpans = false)

            @JvmStatic
            fun builder() = Builder()
        }
    }

    /**
     * Options for configuring metrics.
     *
     * @property enabled Whether to enable metrics.
     */
    data class MetricsApi(val enabled: Boolean = true) {
        companion object {
            @JvmStatic
            fun enabled() = MetricsApi(true)

            @JvmStatic
            fun disabled() = MetricsApi(false)
        }
    }

    /**
     * Options for configuring analytics telemetry. These signals are emitted as
     * OpenTelemetry spans.
     *
     * @property taps If `true`, the plugin publishes a `click` span for each detected tap.
     *   Tap detection itself is governed by [Instrumentations.userTaps]; if that is disabled
     *   no taps are issued and this flag has no effect. Defaults to `true`.
     * @property trackEvents If `true`, the plugin emits a `track` span when a custom
     *   event is tracked (via the LD `afterTrack` hook or [com.launchdarkly.observability.sdk.LDObserve.track]).
     *   Defaults to `true`.
     * @property screenViews If `true`, the plugin emits a `screen_view` span when a screen is shown
     *   (automatically via [Instrumentations.screens] or manually via
     *   [com.launchdarkly.observability.sdk.LDObserve.trackScreenView]). This flag only gates the
     *   span; automatic screen *detection* (and therefore Session Replay `Navigate` events) is
     *   controlled by [Instrumentations.screens]. Defaults to `true`.
     * @property appLifecycle If `true`, the plugin emits app-lifecycle spans as the app moves
     *   between states: `app_foreground` (with `event.lifecycle_state = foreground`) when it enters
     *   the foreground, and `app_background` (with `event.lifecycle_state = background`) when it
     *   enters the background. This flag only gates the span; the matching Session Replay
     *   `Foreground` / `Background` breadcrumbs are emitted regardless. Defaults to `true`.
     */
    data class Analytics(
        val taps: Boolean = true,
        val trackEvents: Boolean = true,
        val screenViews: Boolean = true,
        val appLifecycle: Boolean = true,
        val appLaunch: Boolean = true,
    ) {
        /**
         * Java-friendly fluent builder for [Analytics].
         */
        class Builder {
            private var value = Analytics()

            fun taps(taps: Boolean) = apply { value = value.copy(taps = taps) }
            fun trackEvents(trackEvents: Boolean) = apply { value = value.copy(trackEvents = trackEvents) }
            fun screenViews(screenViews: Boolean) = apply { value = value.copy(screenViews = screenViews) }
            fun appLifecycle(appLifecycle: Boolean) = apply { value = value.copy(appLifecycle = appLifecycle) }
            fun appLaunch(appLaunch: Boolean) = apply { value = value.copy(appLaunch = appLaunch) }

            fun build() = value
        }

        companion object {
            @JvmStatic
            fun builder() = Builder()
        }
    }

    /**
     * This class allows enabling or disabling specific automatic instrumentations.
     *
     * @property crashReporting If `true`, the plugin will automatically report any uncaught exceptions as errors.
     * @property launchTime If `true`, emits launch-time performance telemetry: the legacy TTID/TTFD
     *   histogram metrics, plus the `app.start` span event on `app_launch` (cold/warm via `start.type`,
     *   with `start.duration_ms`). When `false` the `app.start` event is omitted and the `app_launch`
     *   span is anchored at the launch-detection time (rather than back-dated to process start) so it
     *   carries no startup duration. The `app_launch` span itself is still emitted when
     *   [Analytics.appLaunch] is enabled.
     * @property userTaps If `true`, the plugin runs the tap-detection machinery, issuing tap events
     *   from the captured touch stream. Publishing those taps as `click` spans is governed
     *   separately by [Analytics.taps]; Session Replay capture is unaffected by either flag.
     *   Defaults to `true`.
     * @property screens If `true`, the plugin automatically detects screen changes via Android Activity
     *   lifecycle callbacks. This drives the `screen_view` span (gated separately by
     *   [Analytics.screenViews]) and Session Replay `Navigate` events. Defaults to `true`.
     */
    data class Instrumentations(
        val crashReporting: Boolean = true,
        val launchTime: Boolean = false,
        val userTaps: Boolean = true,
        val screens: Boolean = true,
    ) {
        /**
         * Java-friendly fluent builder for [Instrumentations].
         */
        class Builder {
            private var value = Instrumentations()

            fun crashReporting(crashReporting: Boolean) = apply { value = value.copy(crashReporting = crashReporting) }
            fun launchTime(launchTime: Boolean) = apply { value = value.copy(launchTime = launchTime) }
            fun userTaps(userTaps: Boolean) = apply { value = value.copy(userTaps = userTaps) }
            fun screens(screens: Boolean) = apply { value = value.copy(screens = screens) }

            fun build() = value
        }

        companion object {
            @JvmStatic
            fun builder() = Builder()

            /** Every automatic instrumentation enabled. */
            @JvmStatic
            fun enabled() = Instrumentations(
                crashReporting = true,
                launchTime = true,
                userTaps = true,
                screens = true,
            )

            /**
             * Every automatic instrumentation disabled. Note this also turns off user-tap detection
             * (so no `click` spans are emitted regardless of [Analytics.taps]) and automatic screen
             * detection (so no `screen_view` / Session Replay `Navigate` events).
             */
            @JvmStatic
            fun disabled() = Instrumentations(
                crashReporting = false,
                launchTime = false,
                userTaps = false,
                screens = false,
            )
        }
    }

    /**
     * Defines the logging levels for telemetry data. These levels correspond to the OpenTelemetry Log Severity.
     *
     * The levels are ordered by severity, from `TRACE` (least severe) to `FATAL` (most severe).
     * Setting a `logsApiLevel` in [ObservabilityOptions] to a specific level means that
     * logs of that level and all higher severity levels will be exported.
     *
     * For instance, setting the level to `INFO` will cause `INFO`, `WARN`, `ERROR`, and `FATAL`
     * logs (and their variants) to be exported, while `TRACE` and `DEBUG` logs will be ignored.
     *
     * The `NONE` level can be used to disable log exporting entirely.
     *
     * @see <a href="https://opentelemetry.io/docs/specs/otel/logs/data-model/#severity-fields">OpenTelemetry Log Data Model - Severity</a>
     *
     * @property level The integer representation of the log level, as defined by OpenTelemetry.
     */
    enum class LogLevel(val level: Int) {
        TRACE(1),
        TRACE2(2),
        TRACE3(3),
        TRACE4(4),
        DEBUG(5),
        DEBUG2(6),
        DEBUG3(7),
        DEBUG4(8),
        INFO(9),
        INFO2(10),
        INFO3(11),
        INFO4(12),
        WARN(13),
        WARN2(14),
        WARN3(15),
        WARN4(16),
        ERROR(17),
        ERROR2(18),
        ERROR3(19),
        ERROR4(20),
        FATAL(21),
        FATAL2(22),
        FATAL3(23),
        FATAL4(24),
        NONE(Int.MAX_VALUE)
    }

    /**
     * Java-friendly fluent builder for [ObservabilityOptions].
     *
     * Kotlin callers can keep using the [ObservabilityOptions] constructor with named/default
     * arguments. This builder exists so Java callers, which cannot omit Kotlin default parameters,
     * can configure only the options they care about. Every setter defaults to the same value as
     * the [ObservabilityOptions] primary constructor.
     *
     * ```java
     * ObservabilityOptions options = ObservabilityOptions.builder()
     *     .debug(true)
     *     .otlpEndpoint(BuildConfig.OTLP_ENDPOINT)
     *     .instrumentations(
     *         ObservabilityOptions.Instrumentations.builder()
     *             .launchTime(true)
     *             .build())
     *     .build();
     * ```
     */
    class Builder {
        private var options = ObservabilityOptions()

        fun enabled(enabled: Boolean) = apply { options = options.copy(enabled = enabled) }
        fun serviceName(serviceName: String) = apply { options = options.copy(serviceName = serviceName) }
        fun serviceVersion(serviceVersion: String) = apply { options = options.copy(serviceVersion = serviceVersion) }
        fun otlpEndpoint(otlpEndpoint: String) = apply { options = options.copy(otlpEndpoint = otlpEndpoint) }
        fun backendUrl(backendUrl: String) = apply { options = options.copy(backendUrl = backendUrl) }
        fun contextFriendlyName(contextFriendlyName: String?) = apply { options = options.copy(contextFriendlyName = contextFriendlyName) }
        fun resourceAttributes(resourceAttributes: Attributes) = apply { options = options.copy(resourceAttributes = resourceAttributes) }
        fun customHeaders(customHeaders: Map<String, String>) = apply { options = options.copy(customHeaders = customHeaders) }

        /** Sets the session background timeout in milliseconds (Java-friendly overload). */
        fun sessionBackgroundTimeoutMillis(millis: Long) = apply {
            options = options.copy(sessionBackgroundTimeout = millis.milliseconds)
        }
        fun sessionBackgroundTimeout(sessionBackgroundTimeout: Duration) = apply {
            options = options.copy(sessionBackgroundTimeout = sessionBackgroundTimeout)
        }
        fun debug(debug: Boolean) = apply { options = options.copy(debug = debug) }
        fun logsApiLevel(logsApiLevel: LogLevel) = apply { options = options.copy(logsApiLevel = logsApiLevel) }
        fun tracesApi(tracesApi: TracesApi) = apply { options = options.copy(tracesApi = tracesApi) }
        fun metricsApi(metricsApi: MetricsApi) = apply { options = options.copy(metricsApi = metricsApi) }
        fun analytics(analytics: Analytics) = apply { options = options.copy(analytics = analytics) }
        fun instrumentations(instrumentations: Instrumentations) = apply { options = options.copy(instrumentations = instrumentations) }
        fun logAdapter(logAdapter: ObserveLogAdapter) = apply { options = options.copy(logAdapter = logAdapter) }
        fun loggerName(loggerName: String) = apply { options = options.copy(loggerName = loggerName) }
        fun telemetryInspector(telemetryInspector: TelemetryInspector?) = apply { options = options.copy(telemetryInspector = telemetryInspector) }

        fun build() = options
    }

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }
}

File: sdk/@launchdarkly/flutter/packages/observability/lib/src/options/observability_options.dart
// Ported from
// sdk/@launchdarkly/mobile-dotnet/observability/observe/plugin/ObservabilityOptions.cs.

import '../instrumentation/click/ld_click.dart';
import '../plugin/observability_config.dart';

/// Severity threshold for exported logs. Mirrors the `LogLevel` enums in the
/// Android (`ObservabilityOptions.LogLevel`) and iOS (`ObservabilityOptions.LogLevel`)
/// SDKs. The [severity] values match the OpenTelemetry log severity numbers so
/// they can be forwarded across the native bridge unchanged.
///
/// Native-only: the web/Dart pipeline currently ignores this value.
enum ObservabilityLogLevel {
  /// OTel severity `TRACE` (1): export every log.
  trace(1),

  /// OTel severity `TRACE2` (2).
  trace2(2),

  /// OTel severity `TRACE3` (3).
  trace3(3),

  /// OTel severity `TRACE4` (4).
  trace4(4),

  /// OTel severity `DEBUG` (5).
  debug(5),

  /// OTel severity `DEBUG2` (6).
  debug2(6),

  /// OTel severity `DEBUG3` (7).
  debug3(7),

  /// OTel severity `DEBUG4` (8).
  debug4(8),

  /// OTel severity `INFO` (9). The default threshold.
  info(9),

  /// OTel severity `INFO2` (10).
  info2(10),

  /// OTel severity `INFO3` (11).
  info3(11),

  /// OTel severity `INFO4` (12).
  info4(12),

  /// OTel severity `WARN` (13).
  warn(13),

  /// OTel severity `WARN2` (14).
  warn2(14),

  /// OTel severity `WARN3` (15).
  warn3(15),

  /// OTel severity `WARN4` (16).
  warn4(16),

  /// OTel severity `ERROR` (17).
  error(17),

  /// OTel severity `ERROR2` (18).
  error2(18),

  /// OTel severity `ERROR3` (19).
  error3(19),

  /// OTel severity `ERROR4` (20).
  error4(20),

  /// OTel severity `FATAL` (21).
  fatal(21),

  /// OTel severity `FATAL2` (22).
  fatal2(22),

  /// OTel severity `FATAL3` (23).
  fatal3(23),

  /// OTel severity `FATAL4` (24).
  fatal4(24),

  /// Disables log exporting entirely.
  none(0x7fffffff);

  /// The OpenTelemetry log severity number for this level.
  final int severity;

  const ObservabilityLogLevel(this.severity);
}

/// Controls automatic trace generation. Mirrors `TracesApi` (Android) and
/// `AppTracing` (iOS).
///
/// Native-only: the web/Dart pipeline currently ignores these values.
class TracesOptions {
  /// Whether to automatically record errors and exceptions as spans.
  final bool includeErrors;

  /// Whether to automatically record UI performance and other events as spans.
  final bool includeSpans;

  /// Creates trace options; both kinds of automatic spans are on by default.
  const TracesOptions({this.includeErrors = true, this.includeSpans = true});
}

/// Toggles for SDK-side instrumentation. Mirrors `InstrumentationOptions` in
/// the .NET MAUI bridge, extended with the Dart-only [debugPrint] control.
class InstrumentationOptions {
  /// Reserved for network request instrumentation. Currently has no effect:
  /// HTTP requests are not instrumented on any platform, because Flutter's
  /// HTTP clients go through `dart:io`, which native instrumentation cannot
  /// see.
  final bool networkRequests;

  /// Whether to instrument launch times (native bridge only).
  final bool launchTimes;

  /// Whether to automatically report uncaught exceptions as errors. Mirrors
  /// Android `Instrumentations.crashReporting` and iOS `CrashReporting`.
  ///
  /// Native-only. Defaults to `true` to match the native SDK defaults.
  final bool crashReporting;

  /// Controls instrumentation of `debugPrint` in the Dart OpenTelemetry
  /// pipeline. Defaults to [DebugPrintSetting.releaseOnly].
  final DebugPrintSetting debugPrint;

  /// Creates instrumentation options; every instrumentation is on by default,
  /// with `debugPrint` captured in release builds only.
  const InstrumentationOptions({
    this.networkRequests = true,
    this.launchTimes = true,
    this.crashReporting = true,
    this.debugPrint = const DebugPrintReleaseOnly(),
  });
}

/// Analytics telemetry emitted as OpenTelemetry spans. Mirrors Android
/// `ObservabilityOptions.Analytics`.
///
/// [taps] and [views] map to the native `analytics.taps` / `analytics.screenViews`
/// publish gates on both iOS and Android. [trackEvents] gates the `track` span
/// emitted by the cross-platform Dart pipeline (web and mobile) as well as the
/// native Android span.
class AnalyticsOptions {
  /// Whether to capture taps and publish a `click` event for each.
  ///
  /// Flutter draws its whole UI into one native view, so the tapped widget is
  /// resolved in Dart (see `SessionReplayCapture`, which hosts the detector) and
  /// reported to the native SDK; this flag gates that detection as well as the
  /// native publish gates it drives (`analytics.taps` and
  /// `instrumentation.userTaps`). On mobile the `click` span is gated by this
  /// flag while the Session Replay `Click` event is emitted regardless — matching
  /// how [views] treats `Navigate`. Defaults to `true`.
  final bool taps;

  /// Whether to emit spans for screen/page views. Screen views are reported from
  /// Dart (see `LDNavigatorObserver`); on web this gates the Dart `screen_view`
  /// span, and on Android and iOS it maps to the native `analytics.screenViews`
  /// gate. It gates the span only — the mobile Session Replay `Navigate` event is
  /// emitted regardless. Defaults to `true`.
  final bool views;

  /// Whether to emit a `track` span when a custom event is tracked, either
  /// through the LaunchDarkly client's `afterTrack` hook or the manual
  /// `LDObserve.track` API. Defaults to `true`.
  final bool trackEvents;

  /// Whether to emit app-lifecycle spans as the app changes state.
  ///
  /// On every platform this gates the Dart `device.app.lifecycle` span, emitted
  /// on each Flutter `AppLifecycleState` change (`resumed`, `inactive`,
  /// `hidden`, `paused`, `detached`). On iOS and Android it also gates the
  /// native `app_foreground` / `app_background` spans. Defaults to `true`.
  final bool appLifecycle;

  /// Whether to emit an `app_launch` span (carrying `event.launch_type` —
  /// `install` / `update` / `relaunch` — and version fields) once per process
  /// launch. Mobile-only (iOS, Android); a no-op elsewhere. Defaults to `true`.
  final bool appLaunch;

  /// Recognizes the application's own widget types as click targets, so a
  /// design-system `PrimaryButton` reports itself instead of the anonymous
  /// `InkWell` it is built from.
  ///
  /// Registers a type once for the whole app, where the `LDClick` widget names a
  /// single instance. Dart-side only — it runs while the tapped widget is
  /// resolved, so it is not sent to native.
  ///
  /// ```dart
  /// AnalyticsOptions(
  ///   customClickTargetResolver: (widget) => switch (widget) {
  ///     PrimaryButton(:final label) =>
  ///       LDClickTargetInfo(tag: 'PrimaryButton', text: label),
  ///     _ => null,
  ///   },
  /// )
  /// ```
  final LDClickTargetResolver? customClickTargetResolver;

  /// Creates analytics options; all analytics telemetry is on by default.
  const AnalyticsOptions({
    this.taps = true,
    this.views = true,
    this.trackEvents = true,
    this.appLifecycle = true,
    this.appLaunch = true,
    this.customClickTargetResolver,
  });

  /// All analytics telemetry enabled. Convenience mirroring Swift's
  /// `Analytics.enabled`; equivalent to `const AnalyticsOptions()`.
  ///
  /// ```dart
  /// ObservabilityOptions(analytics: AnalyticsOptions.enabled)
  /// ```
  static const AnalyticsOptions enabled = AnalyticsOptions();

  /// All analytics telemetry disabled. Convenience mirroring Swift's
  /// `Analytics.disabled`.
  ///
  /// ```dart
  /// ObservabilityOptions(analytics: AnalyticsOptions.disabled)
  /// ```
  static const AnalyticsOptions disabled = AnalyticsOptions(
    taps: false,
    views: false,
    trackEvents: false,
    appLifecycle: false,
    appLaunch: false,
  );
}

/// Configuration for the LaunchDarkly observability plugin. Field set mirrors
/// `LaunchDarkly.Observability.ObservabilityOptions` in the .NET bridge.
///
/// This type is platform-agnostic on purpose: it carries no transport details
/// (no pigeon wire types), so it can be passed to either the native (mobile)
/// or web implementation. The native wire conversion lives in
/// `platform/io/native_options_codec.dart`.
class ObservabilityOptions {
  /// The [serviceName] used when none is given.
  static const String defaultServiceName = 'observability-flutter';

  /// The LaunchDarkly OTLP collector used when no [otlpEndpoint] is given.
  static const String defaultOtlpEndpoint =
      'https://otel.observability.app.launchdarkly.com:4318';

  /// The LaunchDarkly observability backend used when no [backendUrl] is
  /// given.
  static const String defaultBackendUrl =
      'https://pub.observability.app.launchdarkly.com';

  /// Whether observability telemetry is recorded. Defaults to `true`.
  ///
  /// When `false`, no spans or logs are recorded from Dart: flag evaluations,
  /// `LDObserve.startSpan`, `recordException`, `recordLog`, `debugPrint` and
  /// lifecycle capture are all no-ops, and on web nothing is exported. Session
  /// replay is controlled separately by `SessionReplayOptions.isEnabled`; while
  /// it is on, screen views, clicks, track events and identifies are still
  /// forwarded to native, because the replay timeline is built from them. The
  /// flag is also passed to the native SDK for its automatic instrumentation.
  final bool isEnabled;

  /// The `service.name` resource attribute. Defaults to [defaultServiceName],
  /// which names the SDK rather than your app — set it to identify your app.
  final String serviceName;

  /// The `service.version` resource attribute, such as a release version or
  /// Git SHA.
  ///
  /// When `null` (the default), iOS and Android report the host app's version
  /// (`CFBundleShortVersionString` / `versionName`, which Flutter sets from the
  /// `version` in `pubspec.yaml`), and web omits `service.version`.
  final String? serviceVersion;

  /// The OTLP/HTTP endpoint telemetry is exported to (without the `/v1/...`
  /// signal path). Defaults to [defaultOtlpEndpoint]; override it to send data
  /// through a proxy.
  final String otlpEndpoint;

  /// The LaunchDarkly observability backend used for session and replay
  /// metadata. Defaults to [defaultBackendUrl].
  final String backendUrl;

  /// A human-readable name for the current context, shown in the
  /// observability UI instead of its key.
  final String? contextFriendlyName;

  /// Extra OTel Resource attributes attached to every exported signal on
  /// every platform, such as `deployment.environment`.
  ///
  /// Use strings, numbers, booleans, or homogeneous lists of those; other
  /// values are not portable across platforms. Use [serviceName] and
  /// [serviceVersion] rather than the `service.*` keys, which the SDK sets
  /// itself.
  final Map<String, Object?>? attributes;

  /// Extra HTTP headers added to OTLP exports (e.g. for proxies or auth).
  /// Mirrors Android/iOS `customHeaders`. Native-only.
  final Map<String, String> customHeaders;

  /// How long the app may stay in the background before the current session is
  /// ended. Mirrors Android/iOS `sessionBackgroundTimeout`. Native-only.
  /// Defaults to 15 minutes.
  final Duration sessionBackgroundTimeout;

  /// Minimum severity of logs forwarded to the OpenTelemetry logs pipeline.
  /// Use [ObservabilityLogLevel.none] to disable logs. Mirrors Android/iOS
  /// `logsApiLevel`. Native-only. Defaults to [ObservabilityLogLevel.info].
  final ObservabilityLogLevel logsApiLevel;

  /// Controls automatic trace generation. Mirrors Android `tracesApi` and iOS
  /// `tracesApi`. Native-only.
  final TracesOptions traces;

  /// Whether to export metrics. Mirrors Android `metricsApi` and iOS
  /// `metricsApi`. Native-only. Defaults to `true`.
  final bool metricsEnabled;

  /// Analytics telemetry configuration. Mirrors Android
  /// `ObservabilityOptions.analytics`. See each [AnalyticsOptions] field for
  /// which platforms it applies to.
  final AnalyticsOptions analytics;

  /// Toggles for automatic instrumentation: network requests, launch times,
  /// crash reporting and `debugPrint` capture.
  final InstrumentationOptions instrumentation;

  /// Creates observability options. Every argument is optional; the defaults
  /// enable all telemetry and export to LaunchDarkly.
  const ObservabilityOptions({
    this.isEnabled = true,
    this.serviceName = defaultServiceName,
    this.serviceVersion,
    String? otlpEndpoint,
    String? backendUrl,
    this.contextFriendlyName,
    this.attributes,
    this.customHeaders = const {},
    this.sessionBackgroundTimeout = const Duration(minutes: 15),
    this.logsApiLevel = ObservabilityLogLevel.info,
    this.traces = const TracesOptions(),
    this.metricsEnabled = true,
    this.analytics = const AnalyticsOptions(),
    this.instrumentation = const InstrumentationOptions(),
  }) : otlpEndpoint = otlpEndpoint ?? defaultOtlpEndpoint,
       backendUrl = backendUrl ?? defaultBackendUrl;
}

/// Not for export: a public `copyWith` could not clear nullable fields
/// without a sentinel design, so the SDK keeps its one copy internal.
extension ObservabilityOptionsCopy on ObservabilityOptions {
  /// A copy with [attributes] replaced (`null` clears them).
  ObservabilityOptions withAttributes(Map<String, Object?>? attributes) =>
      ObservabilityOptions(
        isEnabled: isEnabled,
        serviceName: serviceName,
        serviceVersion: serviceVersion,
        otlpEndpoint: otlpEndpoint,
        backendUrl: backendUrl,
        contextFriendlyName: contextFriendlyName,
        attributes: attributes,
        customHeaders: customHeaders,
        sessionBackgroundTimeout: sessionBackgroundTimeout,
        logsApiLevel: logsApiLevel,
        traces: traces,
        metricsEnabled: metricsEnabled,
        analytics: analytics,
        instrumentation: instrumentation,
      );
}
